MetaMask/metamask-extension
# Persist state in metaRPCHandler so that we are sure state is persisted before sending back response to actions

## The Issue
While digging around to find the root cause of #15914 and #15918  @jpuri and I came across a [long standing issue](https://consensys.slack.com/archives/C02DUDT2WTU/p1663879099541579?thread_ts=1663878724.445999&cid=C02DUDT2WTU) that is now causing new problems in the context of MV3:

Because the browser storage API is asynchronous we cannot make guarantees about when our calls to [persist background store data](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/background.js#L387) will resolve, but to avoid laggy UX for any action that requires data persistence we optimistically reflect changes in the UI as if the persistence has completed. We believe that even within MV2 we may be vulnerable to data corruption issues with our current implementation that arise if the background process is disrupted while it is persisting data across multiple calls -  some of a batch of related data might get saved while other bits of it are not causing corruption. 

Because in MV3 the background process (now a service worker) is constantly (and unpredictably) winding down, the async storage API is leading to another issue: there is a small window of time when, if the service worker's lifecycle ends immediately after a user action and [the action is executed](https://github.com/MetaMask/metamask-extension/blob/develop/ui/store/action-queue/index.js#L62) in the background but not yet persisted, the action will [not be added to the replay queue](https://github.com/MetaMask/metamask-extension/blob/develop/ui/store/action-queue/index.js#L64) since the action was in fact executed - just not yet persisted.

## The Solution
This solution we've landed on is to persist the data that results from background state changes further upstream than previously, in the controller multiplex. This allows us to persist the data before the action is resolved ensuring that the queuing system will work - [if the action promise resolves](https://github.com/MetaMask/metamask-extension/blob/24a3156a2ba9e7ad55c44e3b5aa655d99bfbb500/ui/store/action-queue/index.js#L140) we can guarantee it has been persisted, if an error is thrown [we can push it to the queue and retry](https://github.com/MetaMask/metamask-extension/blob/24a3156a2ba9e7ad55c44e3b5aa655d99bfbb500/ui/store/action-queue/index.js#L64).

## More Information

* Fixes #15914 & #15918
* See also: [slack thread](https://consensys.slack.com/archives/C02DUDT2WTU/p1663878724445999)

### Before



### After

- The service worker dying (or being terminated) shortly after removing a network doesn't see the network deletion failing to persist (fixes https://github.com/MetaMask/metamask-extension/issues/15918):
  - https://user-images.githubusercontent.com/34557516/192581100-67bc1bab-e4ea-4231-8e00-f84ef54abbe8.mov

- The service worker dying (or being terminated) shortly after removing a contact doesn't see the contact deletion failing to persist (fixes https://github.com/MetaMask/metamask-extension/issues/15914):
  - https://user-images.githubusercontent.com/34557516/192581844-a654cfc8-0703-4851-a1b1-2123b1c877ac.mov

- The service worker dying (or being terminated) shortly after confirming a transaction doesn't lead to the extension failing to bump nonce or persist the state that recognizes the transaction is in flight. (fixes issue identified by Mariona [here](https://consensys.slack.com/archives/C02DUDT2WTU/p1663929271170509?thread_ts=1663878724.445999&cid=C02DUDT2WTU) - [video](https://consensys.slack.com/files/U032XS4NP9U/F043F2HP5V4/submit-tx.webm)):
  - https://user-images.githubusercontent.com/34557516/192584838-7c18af26-9202-419f-809b-fec81cc1448d.mov





## Manual Testing Steps

The most concerning risk with this PR is that we break state persistence somehow on the MV2 build. When testing, we'll want to ensure state is loaded from disk correctly and saved back to disk correctly.

To test loading, you can follow the same steps as for testing an extension update (i.e. start with a different build, switch to this one, and start the extension).

To test state saving, make any changes in the wallet that would be saved and restart it, then check the extension to ensure the changes remain.

I don't know of a great way to directly test the MV3 changes because they resulted in intermittent failures.

## Pre-Merge Checklist


### + If there are functional changes:
